samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,180 @@
+from pathlib import Path
+from typing import Optional
+
+from samcli.lib.utils import osutils
+from samcli.lib.utils.osutils import rmtree_callback
+
+LOG = logging.getLogger(__name__)
+
+
+class CloneRepoException(Exception):
+    """
+    Exception class when clone repo fails.
+    """
+
+
+class CloneRepoUnstableStateException(CloneRepoException):
+    """
+    Exception class when clone repo enters an unstable state.
+    """
+
+
+class GitRepo:
+    """
+    Class for managing a Git repo, currently it has a clone functionality only
+
+    Attributes
+    ----------
+    _url: str
+        The URL of this Git repository, example "https://github.com/aws/aws-sam-cli"
+    _name: str
+        The name of this Git repository, example "aws-sam-cli"
+    _local_path: Path
+        The path of the local clone of this Git repository
+    _clone_attempted: bool
+        whether an attempt to clone this Git repository took place or not
+
+    Methods
+    -------
+    clone(self, clone_dir: Path, clone_name=None, replace_existing=False) -> Path:
+        creates a local clone of this Git repository
+    """
+
+    def __init__(self, url: str, name: str) -> None:
+        self._url: str = url
+        self._name: str = name
+        self._local_path: Optional[Path] = None
+        self._clone_attempted: bool = False
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def local_path(self) -> Optional[Path]:
+        return self._local_path
+
+    @local_path.setter
+    def local_path(self, value: Path) -> None:
+        self._local_path = value
+
+    @property
+    def clone_attempted(self) -> bool:
+        return self._clone_attempted
+
+    @clone_attempted.setter
+    def clone_attempted(self, value: bool) -> None:
+        self._clone_attempted = value
+
+    @staticmethod
+    def _ensure_clone_directory_exists(clone_dir: Path) -> None:
+        try:
+            clone_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+        except OSError as ex:
+            LOG.warning("WARN: Unable to create clone directory.", exc_info=ex)
+            raise
+
+    @staticmethod
+    def _git_executable() -> str:
+        if platform.system().lower() == "windows":
+            executables = ["git", "git.cmd", "git.exe", "git.bat"]
+        else:
+            executables = ["git"]
+
+        for executable in executables:
+            try:
+                subprocess.Popen([executable], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                # No exception. Let's pick this
+                return executable
+            except OSError as ex:
+                LOG.debug("Unable to find executable %s", executable, exc_info=ex)
+
+        raise OSError("Cannot find git, was looking at executables: {}".format(executables))
+
+    def clone(self, clone_dir: Path, clone_name: Optional[str] = None, replace_existing: bool = False) -> Path:
```

Comment:
If the sole purpose is to clone a Git repository, what is the reasoning used for which arguments are in the constructor, and which are in `clone`?

Reply:
my intention is to make `clone` the initial functionality but not the only one. that is why in the docstring I added the word "currently":

> Class for managing a Git repo, **currently** it has a clone functionality only

Reply:
Why there is a `clone_name`, the only usage is `self._git_repo.clone(clone_dir=shared_dir, replace_existing=True)` and it does not give `clone_name` any value.

Reply:
`clone_name` is unnecessary. It is being used at all.

Reply:
I still don't understand what value all those names bring. A Git directory (URL) is cloned to a directory (name). What other "name" is needed?

Reply:
Actually it's more confusing on second look. There's `_name`, `_local_path`, `clone_name` and `clone_dir`.

`git clone` takes a single optional name as argument (the clone directory); why is there so much complexity here? I understand this is mostly existing logic, but it feels like a great potential win to significantly simplify this.

The whole `_persist_local_repo` logic seems unnecessary as well. Why not directly clone to the target directory?

Reply:
`persist_local_repo` is to move the created clone from temp to its final destination. It should be clear why it is created at temp in the first place, right?


why are you mixing method parameters; `clone_name` and `clone_dir` with class attributes?
For the class attributes:
* A GitRepo **has a** _url  # example: "https://github.com/aws/aws-sam-cli-app-pipeline-templates"
* A GitRepo **has a** _name # example "aws-sam-cli-app-pipeline-templates"
* A GitRepo **has a** _local_path # example "~/.aws-sam/aws-sam-cli-app-pipeline-templates"

* _local_path is set only after clone as it is derived from clone's destination `{clone_dir}/{clone_name}`